frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -811,8 +812,10 @@ TypeInference::assignment(const IR::Node* errorPosition, const IR::Type* destTyp
             }
             if (!changes)
                 vec = si->components;
-            if (initType->is<IR::Type_UnknownStruct>() || changes)
+            if (initType->is<IR::Type_UnknownStruct>() || changes) {
                 sourceExpression = new IR::StructExpression(type, type, vec);
+                setType(sourceExpression, destType);
```

Comment:
@mbudiu-vmw can you explain what this block is doing?

Reply:
This is one of these cases where a type has been inferred for a StructExpression used in the RHS of an assignment.
A struct expression written without a type, e.g. `{a = 1}` has type `Type_UnknownStruct`. If the type is inferred it is written like `(S) {a = 1}` and the type becomes `Type_Struct`. In this case the type has been inferred from the LHS of the assignment (that's why it's `destType`).

Notice that the situation handled in this block is more complicated, since you can have something like:
`S a = (S){ a = { b = 1 }}` when `S` is a nested struct. So, even if the outer struct has the correct type `S`, the inner one does still have a `Type_UnknownStruct`, so the code prior to this has to recurse over the struct expression to make sure that all its fields are correct as well. Their types are inferred from the fields of `S`.

Reply:
A similar inference pattern is used when structs are compared.

Reply:
This code also handles mixed assignments that use list expressions, casting them to StructExpressions: e.g.:
`S s = { a = { 1 } }`